src/server/spatial_map.h

```diff
@@ -0,0 +1,83 @@
+/*
+Minetest
+Copyright (C) 2024, ExeVirus <nodecastmt@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+
+#include <functional>
+#include <vector>
+#include <unordered_set>
+#include "irrlichttypes_bloated.h"
+#include "constants.h"
+
+namespace server
+{
+class SpatialMap
+{
+public:
+	void insert(u16 id, const v3f &pos, bool postIteration = false);
+	void remove(u16 id, const v3f &pos, bool postIteration = false);
+	void remove(u16 id);
+	void removeAll();
+	void updatePosition(u16 id, const v3f &oldPos, const v3f &newPos);
+	void getRelevantObjectIds(const aabb3f &box, const std::function<void(u16 id)> &callback);
+	void handleInsertsAndDeletes();
+
+protected:
+	struct SpatialKey {
+		u16 padding_or_optional_id{0};
+		s16 x;
+		s16 y;
+		s16 z;
+
+		SpatialKey(s16 _x, s16 _y, s16 _z, bool _shrink = true) {
+			if(_shrink) {
+				x = _x >> 4;
+				y = _y >> 4;
+				z = _z >> 4;
+			} else {
+				x = _x;
+				y = _y;
+				z = _z;
+			}
+		}
+		SpatialKey(const v3f &_pos) : SpatialKey(_pos.X / BS, _pos.Y / BS, _pos.Z / BS){}
+		// The following use case is for storing pending insertions and deletions while iterating
+		// using the extra 16 bit padding makes keeping track of them super efficient for hashing.
+		SpatialKey(const v3f &_pos, const u16 id) : SpatialKey(_pos.X / BS, _pos.Y / BS, _pos.Z / BS, false){
+			padding_or_optional_id = id;
+		}
+
+		bool operator==(const SpatialKey &other) const {
+			return (x == other.x && y == other.y && z == other.z);
+		}
+	};
+
+	struct SpatialKeyHash {
+		auto operator()(const SpatialKey &key) const -> size_t {
+			return std::hash<size_t>()(*reinterpret_cast<const size_t*>(&key));
+		}
+	};
+
+	std::unordered_multimap<SpatialKey, u16, SpatialKeyHash> m_cached;
+	std::unordered_set<SpatialKey, SpatialKeyHash> m_pending_inserts;
+	std::unordered_set<SpatialKey, SpatialKeyHash> m_pending_deletes;
```

Comment:
this seems precarious, if more than one object in the same map block is added or deleted it could be ignored (see my note about undefined behavior above). Why not just use a vector?

Reply:
I keep them in unordered_sets because there are cases where I might add and delete the same entity in one loop, and I want to ensure that entity is properly (not) created. 

But yes the more I think about it, you are totally correct and I will be taking the optional padding out of the equation and use a separate type with better naming to make this all much more readable and safe.

Reply:
I'm not sure I understand, aren't all pending inserts applied after all pending deletes?

So if an entity is added in a loop and then deleted in that same loop iteration, won't it incorrectly end up not deleted?

That might just be considered an error on the part of the API user, I think that would be an acceptable tradeoff, but entity updates are also split into a remove and insert and added to the queued lists.
Updating an entity while iterating before deleting it seems like a more legitimate way to use the API. Or at least easier to do on accident. Could imagine some code that calculates an updated object position before checking whatever it's expired and needs to be deleted.

I maybe could use an event-style queue or vector (whichever is faster) with an event type indicating whatever it's a deletion or insertion since that would trivially be ordered. Would I think only end up with one byte of padding. Does mean a insert follow by a delete would be a lot more expensive than it needs to be.

Reply:
Had more time to think about this. It's messy in my current solution agreed.

I have decided to use a PendingOperation struct holding all the info (and whether it's being inserted or deleted), and go with a vector solution as you suggested. I just don't think there's a better way. 

In worst cases with a vector, you might get an entity who has been requested to be removed multiple times, which will certainly hurt performance (cause we'll iterate the whole map to try to delete it the second+ times), but it's the only sane way to do things....